File: sem3/map/Lab/3/toylanguage/src/main/java/com/company/Models/Statements/OpenFileStatement.java
package com.company.Models.Statements;

import com.company.Models.*;
import com.company.Models.ProgramState.ProgramState;
import com.company.Utils.MyIDictionary;
import com.company.Utils.Pair;

import java.io.*;

public class OpenFileStatement implements IStatement {
    private String variableFileID;
    private String filePath;

    public OpenFileStatement(String variableFileID, String filePath) {
        this.variableFileID = variableFileID;
        this.filePath = filePath;
    }

    @Override
    public ProgramState execute(ProgramState state) throws Exception {
        MyIDictionary<Integer, Pair<String, BufferedReader>> fileTable = state.getFileTable();
        MyIDictionary<String, Integer> symbolTable = state.getSymTable();

        for(Integer key: fileTable.getKeys()) {
            if (fileTable.getValue(key).getFirst().equals(this.filePath)) {
                throw ExceptionFactory.create("file", "File is already open!");
            }
        }

        BufferedReader reader;

        try {
            reader = new BufferedReader(new FileReader(this.filePath));
        } catch (FileNotFoundException e) {
            throw ExceptionFactory.create("file", "Couldn't open file!");
        }

        int fileID = reader.hashCode();

        fileTable.add(fileID, new Pair<>(this.filePath, reader));
        symbolTable.add(this.variableFileID, fileID);

        return state;
    }

    @Override
    public String toString() {
        return "OpenFile(" + this.variableFileID + ", \"" + this.filePath + "\")";
    }
}

File: sem3/map/Lab/toylanguage_C#/ToyLanguage/ToyLanguage/Models/Statements/OpenFileStatement.cs
using System;
using System.IO;
using ToyLanguage.Models.Interfaces;
using ToyLanguage.Models.States;

namespace ToyLanguage.Models.Statements
{
    public class OpenFileStatement : IStatement
    {
        private readonly string _variableFileID;
        private readonly string _filePath;

        public OpenFileStatement(string variableFileId, string filePath)
        {
            _variableFileID = variableFileId;
            _filePath = filePath;
        }
        
        public ProgramState Execute(ProgramState state)
        {
            StreamReader fileStream;

            try
            {
                fileStream = new StreamReader(_filePath);
            }
            catch (Exception e)
            {
                throw new Exception($"Couldn\'t open file with path: {_filePath}");
            }

            int fileID = fileStream.GetHashCode();
            
            state.FileTable[fileID] = new Tuple<string, StreamReader>(_filePath, fileStream);
            state.SymbolTable[_variableFileID] = fileID;
            
            return state;
        }

        public override string ToString()
        {
            return $"OpenFile({_variableFileID}, {_filePath})";
        }
    }
}